spec/sane_patch_spec.rb

```diff
@@ -2,4 +2,108 @@
   it "has a version number" do
     expect(SanePatch::VERSION).not_to be nil
   end
+
+  describe ".patch" do
+    let(:gem_name) { "some_gem" }
+    let(:installed_gem_version) { nil }
+
+    before do
+      if installed_gem_version
+        allow(Gem.loaded_specs).to receive(:[]).with(gem_name).and_return(
+          double("Fake Gemspec", version: Gem::Version.new(installed_gem_version))
+        )
+      end
+    end
+
+    context "when gem is not installed" do
+      it "raises error for non-existing gem" do
+        expect do
+          ::SanePatch.patch(gem_name, "0.0.1")
+        end.to raise_error(::SanePatch::Errors::GemAbsent)
+      end
+    end
+
+    context "when gem is installed" do
+      let(:installed_gem_version) { "1.0.1" }
+
+      it "raises error for incompatible version" do
+        yielded = false
+
+        aggregate_failures do
+          expect do
+            SanePatch.patch(gem_name, "0.0.1") do
+              yielded = true
+            end
+          end.to raise_error(::SanePatch::Errors::IncompatibleVersion)
+          expect(yielded).to be_falsey
+        end
+      end
+
+      it "raises error for incompatible version constraint" do
+        yielded = false
+
+        aggregate_failures do
+          expect do
+            SanePatch.patch(gem_name, "~> 0.0.1") do
+              yielded = true
+            end
+          end.to raise_error(::SanePatch::Errors::IncompatibleVersion)
+          expect(yielded).to be_falsey
+        end
+      end
+
+      it "raises error for incompatible version constraints" do
+        yielded = false
+
+        aggregate_failures do
+          expect do
+            SanePatch.patch(gem_name, ">= 0.0.1", "< 1.0.0") do
+              yielded = true
+            end
+          end.to raise_error(::SanePatch::Errors::IncompatibleVersion)
```

Comment:
Could we use the [yield-matcher](https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/yield-matchers) here?

It might make sense to have a custom matcher that checks if the guard engaged or not.